File: Diferenciacion/python/extrRich.py
import math

def extrRich(xi, f, h, O):
    #Implementación del derivada por extrapolación de Richardson. 
    #xi es el vector de puntos, deben estar equiespaciados.
    #f es la función a derivar.
    #h es el paso que queremos aplicar.
    #orden es el orden de error que queremos cometer.
    import math

    if (O == 1):
        return([(1/h)*(f(x+h)-f(x)) for x in xi ]) 
    elif (O == 2):
        return([(1/h)*(-f(x+h)+4*f(x+(h/2))-3*f(x)) for x in xi ])
    elif (O == 3):
        return([(1/h)*(f(x+h)-12*f(x+(h/2))+32*f(x+(4/h))-21*f(x)) for x in xi] )
    else:
        print("No se ha programado para mayor que orden 3")

print(extrRich([0,0.25,0.5,0.75,1], lambda x : x**2*math.exp(-x), h=0.25, O = 1))

File: Diferenciacion/R/extrRich.r
extrRich <- function(xi, f, h, O){
#Implementación del derivada por extrapolación de Richardson. 
#xi es el vector de puntos, deben estar equiespaciados.
#f es la función a derivar.
#h es el paso que queremos aplicar.
#orden es el orden de error que queremos cometer.

if (O == 1){
    return(df = (1/h)*(f(xi+h)-f(xi)))
    }  else if (O == 2){
    return(df = (1/h)*(-f(xi+h)+4*f(xi+(h/2))-3*f(xi)))
    } else if (O == 3){
    return(df = (1/h)*(f(xi+h)-12*f(xi+(h/2))+32*f(xi+(4/h))-21*f(xi)))
    } else {
        print("No se ha programado para mayor que orden 3")
    }

}

extrRich(c(0,0.25,0.5,0.75,1), function(x) x**2*exp(-x), h=0.25, O = 1)
